### ISSUE TITLE
Calculate add/sub gradient

### ISSUE THREAD

**Issue by shanegibbs** [raskr/rust-autograd#3]
I have been tinkering together a neural network. However, I have been running into some panics. I have managed to boil it down to calculating the gradient of `ag::add` or `ag::sub`. Here is the boiled down example:

```rust
use ag;
use ndarray::prelude::*;

let x_val = array![[0.1]].into_dyn();
let y_val = array![[0.2]].into_dyn();
let ref x = ag::placeholder(&[-1, 1]);
let ref y = ag::placeholder(&[-1, 1]);
let ref feeds = [(x, &x_val), (y, &y_val)];

let ref z = ag::add(x, y);
println!("{}", z.eval(feeds));

let ref grads = ag::grad(&[z], &[y]);
println!("{}", grads[0].eval(feeds)); // panic here
```

I get the following when I try to run the above:

```
[[0.3]]
thread 'main' panicked at 'ndarray: could not broadcast array from shape: [0] to: [2]', /home/sgibbs/.cargo/registry/src/github.com-1ecc6299db9ec823/ndarray-0.10.14/src/lib.rs:741:13
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

Am I doing something wrong?

Thanks

**Comment by raskr** [comment]
No, that is exactly a bug (thank you for the report).
If a differentiable operation is evaluated successfully, its output's gradient must also be computed successfully. I will investigate the cause later.
Actually, I often suffer from broadcasting bugs in gradient computation...

**Comment by raskr** [comment]
Correction: That is **not** a bug.

I'm sorry, `ag::grad()` can differentiate **scalar** like tensor... (mentioned in https://docs.rs/autograd/0.6.5/autograd/ops/fn.grad.html)

This is the similar spec as Theano's.
Workarounds:

* Simply you can `reduce_sum()` the `z` into a scalar, which is fed to first argument of `ag::grad`.
```rust
    let ref z = ag::add(x, y);
    let ref z = ag::reduce_sum(z, &[0, 1], false);  // insert this line
    let ref grads = ag::grad(&[z], &[y]);
```
* You can use `ag::grad_with_default`. (as mentioned in the link above)

For now, code that violating this rule causes *undefined* behavior like you reported...
This is one of the most serious problem in this library, so searching the solution.

**Comment by shanegibbs** [comment]
Oh yes, I see now. Should that also work for `reduce_mean()` as well as `reduce_sum()`? I seem to get the same broadcast error with `reduce_mean()`.

For example

```rust
let ref z = ag::reduce_mean(z, &[0], false);
```

**Comment by raskr** [comment]
Yes, it should also work.
I'm fixing those problems now, wait a moment..

**Comment by raskr** [comment]
Probably fixed in master head and v0.6.7 👍 
I made overall fixes of reductions and scalar tensor treatments.
Now, this works:
```rust
let ref v = ag::variable(ndarray::arr2(&[[2., 3., 4.]]));
// shape of `z` is [] (zero-ranked)
let ref z = ag::reduce_mean(v, &[0, 1], /*keep_dims=*/false);
let ref g = ag::grad(&[z], &[v])[0];
println!("{}", g.eval(&[]));  //=> [[0.33333334, 0.33333334, 0.33333334]]
```
